controllers/openstackfloatingippool_controller.go

```diff
@@ -0,0 +1,366 @@
+			if err != nil {
+				return ctrl.Result{}, err
+			}
+
+			ipAddress := &ipamv1.IPAddress{
+				ObjectMeta: ctrl.ObjectMeta{
+					Name:      claim.Name,
+					Namespace: claim.Namespace,
+					OwnerReferences: []metav1.OwnerReference{
+						{
+							APIVersion: claim.APIVersion,
+							Kind:       claim.Kind,
+							Name:       claim.Name,
+							UID:        claim.UID,
+						},
+					},
+				},
+				Spec: ipamv1.IPAddressSpec{
+					ClaimRef: corev1.LocalObjectReference{
+						Name: claim.Name,
+					},
+					PoolRef: corev1.TypedLocalObjectReference{
+						APIGroup: pointer.String(infrav1.GroupVersion.Group),
+						Kind:     pool.Kind,
+						Name:     pool.Name,
+					},
+					Address: ip,
+				},
+			}
+
+			if !contains(pool.Spec.PreAllocatedFloatingIPs, ip) && pool.Spec.ReclaimPolicy == infrav1.ReclaimDelete {
+				controllerutil.AddFinalizer(ipAddress, infrav1.DeleteFloatingIPFinalizer)
+			}
+
+			if err = r.Client.Create(ctx, ipAddress); err != nil {
+				return ctrl.Result{}, err
+			}
+
+			claim.Status.AddressRef.Name = ipAddress.Name
+			if err = r.Client.Status().Update(ctx, &claim); err != nil {
+				log.Error(err, "Failed to update IPAddressClaim status", "claim", claim.Name, "ip", ip)
+				return ctrl.Result{}, err
+			}
+			scope.Logger().Info("Claimed IP", "ip", ip)
+		}
+	}
+	if err = r.setIPStatuses(ctx, scope, pool); err != nil {
+		return ctrl.Result{}, err
+	}
+	return ctrl.Result{}, nil
+}
+
+func (r *OpenStackFloatingIPPoolReconciler) reconcileDelete(ctx context.Context, pool *infrav1.OpenStackFloatingIPPool) (ctrl.Result, error) {
+	log := ctrl.LoggerFrom(ctx)
+	ipAddresses := &ipamv1.IPAddressList{}
+	if err := r.Client.List(ctx, ipAddresses, client.InNamespace(pool.Namespace), client.MatchingFields{infrav1.OpenStackFloatingIPPoolNameIndex: pool.Name}); err != nil {
+		return ctrl.Result{}, err
+	}
+	// If there are still IPAddress objects, they might need the pool to be available for deletion
+	if len(ipAddresses.Items) > 0 {
+		log.Info("Waiting for IPAddress to be deleted before deleting OpenStackFloatingIPPool")
+		return ctrl.Result{Requeue: true}, nil
+	}
+	if controllerutil.RemoveFinalizer(pool, infrav1.OpenStackFloatingIPPoolFinalizer) {
+		log.Info("Removing finalizer from OpenStackFloatingIPPool")
+		return ctrl.Result{Requeue: true}, r.Client.Update(context.Background(), pool)
+	}
+	return ctrl.Result{}, nil
+}
+
+func (r *OpenStackFloatingIPPoolReconciler) setIPStatuses(ctx context.Context, scope scope.Scope, pool *infrav1.OpenStackFloatingIPPool) error {
+	ipAddresses := &ipamv1.IPAddressList{}
+	if err := r.Client.List(ctx, ipAddresses, client.InNamespace(pool.Namespace), client.MatchingFields{infrav1.OpenStackFloatingIPPoolNameIndex: pool.Name}); err != nil {
+		return err
+	}
+
+	networkingService, err := networking.NewService(scope)
+	if err != nil {
+		return err
+	}
+
+	floatingIPs, err := networkingService.GetAllFloatingIPs()
+	if err != nil {
+		return err
+	}
```

Comment:
Agreed that every reconcile is way too much, this was fix after testing and accidentally manually deallocating ips the pool thought was available.

The problem with checking at allocation is, that it would only be a real problem when it's in AvailableIPs and as long as it has AvailableIPs it would use them before allocating new one I would be fine with just checking this when fullfilling a claim with an IP from AvailableIPs, the only downside would be that available ips would not update automatically if they were removed from openstack until someone try to claim them.

But then again this entire check might be overkill, if you delete ips allocated by the pool maybe you shouldn't be surprised if you get broken ips.